src/ray/rpc/retryable_grpc_client.cc

```diff
@@ -0,0 +1,157 @@
+        "Timed out while waiting for %s to become available.", server_name_)));
+    pending_requests_bytes_ -= iter->second->GetRequestBytes();
+    pending_requests_.erase(iter);
+  }
+
+  if (pending_requests_.empty()) {
+    server_unavailable_timeout_time_ = std::nullopt;
+    return;
+  }
+
+  RAY_CHECK(server_unavailable_timeout_time_.has_value());
+
+  auto status = channel_->GetState(false);
+  // https://grpc.github.io/grpc/core/md_doc_connectivity-semantics-and-api.html
+  // https://grpc.github.io/grpc/core/connectivity__state_8h_source.html
+  if (status != GRPC_CHANNEL_READY) {
+    RAY_LOG(DEBUG) << "GRPC channel status: " << status;
+  }
+
+  switch (status) {
+  case GRPC_CHANNEL_TRANSIENT_FAILURE:
+  case GRPC_CHANNEL_CONNECTING: {
+    if (server_unavailable_timeout_time_ < absl::Now()) {
+      RAY_LOG(WARNING) << server_name_ << " has been unavailable for more than "
+                       << server_unavailable_timeout_seconds_ << " seconds";
+      server_unavailable_timeout_callback_();
+      // Reset the unavailable timeout.
+      server_unavailable_timeout_time_ =
+          absl::Now() + absl::Seconds(server_unavailable_timeout_seconds_);
+    }
+
+    if (reset_timer) {
+      SetupCheckTimer();
+    }
+
+    break;
+  }
+  case GRPC_CHANNEL_SHUTDOWN: {
+    RAY_LOG(FATAL) << "Channel shoud never go to this status.";
+    break;
+  }
+  case GRPC_CHANNEL_READY:
+  case GRPC_CHANNEL_IDLE: {
+    server_unavailable_timeout_time_ = std::nullopt;
+    // Retry the ones queued.
+    while (!pending_requests_.empty()) {
+      pending_requests_.begin()->second->CallMethod();
+      pending_requests_.erase(pending_requests_.begin());
+    }
+    pending_requests_bytes_ = 0;
+    break;
+  }
+  default: {
+    RAY_LOG(FATAL) << "Not covered status: " << status;
+  }
+  }
+}
+
+void RetryableGrpcClient::Retry(std::shared_ptr<RetryableGrpcRequest> request) {
+  // In case of transient network error, we queue the request and these requests
+  // will be executed once network is recovered.
+  auto request_bytes = request->GetRequestBytes();
+  auto self = shared_from_this();
+  if (pending_requests_bytes_ + request_bytes > max_pending_requests_bytes_) {
+    RAY_LOG(WARNING) << "Pending queue for failed request has reached the "
+                     << "limit. Blocking the current thread until network is recovered";
+    if (!server_unavailable_timeout_time_.has_value()) {
+      server_unavailable_timeout_time_ =
+          absl::Now() + absl::Seconds(server_unavailable_timeout_seconds_);
+    }
+    while (server_unavailable_timeout_time_.has_value()) {
+      std::this_thread::sleep_for(
+          std::chrono::milliseconds(check_channel_status_interval_milliseconds_));
+
+      if (self.use_count() == 2) {
+        // This means there are no external owners of this client
+        // and the client is considered shut down.
+        // The only two internal owners are caller of Retry and self.
```

Comment:
curious how do we make sure who-owns-retry-client still holds the client?

Reply:
Normally, it's stored in the class member field.

Reply:
> The only two internal owners are caller of Retry

Do you mean the caller for retry function?